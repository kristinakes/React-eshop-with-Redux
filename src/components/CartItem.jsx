import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { TrashIcon, MinusIcon, PlusIcon } from "@heroicons/react/24/outline";

import { cartActions } from "../store/cartSlice";

function CartItem(props) {
  const dispatch = useDispatch();
  const { id, title, quantity, totalPrice, year, ram, warranty, price } =
    props.item;

  const imageName = title.replace(" ", "_").toLowerCase();
   const image = `/assets/product_images/${imageName}.png`;

  const removeItemFromCartHandler = () => {
    dispatch(cartActions.updateItemInCart(id));
  };

  const addItemToCartHAndler = () => {
    dispatch(
      cartActions.addItemToCart({
        id,
        title,
        totalPrice,
        year,
        ram,
        warranty,
        price,
      })
    );
  };

  return (
    <div className="grid grid-cols-2 gap-4 min-w-fit justify-center content-center sm:grid-cols-4 sm:gap-2 shadow-md m-5 p-3">
      <div className="size-48 m-auto">
        <Link to={`/Product/${id}`}>
          <img
            className="object-fill scale-75 z-0"
            src={image}
            alt={title}
          />
        </Link>
      </div>
      <div>
        <Link to={`/Product/${id}`}>
          <h3 className="font-bold text-xl text-blue my-5">{title}</h3>
        </Link>
        <div className="font-semibold flex flex-wrap gap-3 md:flex-col md:gap-1">
          <p>Year: {year}</p>
          <p>RAM memory: {ram}</p>
          <p>Warranty: {warranty}</p>
          <p>Price: {price}</p>
        </div>
      </div>
      <div className="flex justify-center items-center">
        <button id="remove-item-btn" className="btn-blue" onClick={removeItemFromCartHandler}>
          {quantity > 1 && <MinusIcon className="h-6 w-6" aria-hidden="true" />}
          {quantity === 1 && (
            <TrashIcon className="h-6 w-6" aria-hidden="true" />
          )}
        </button>
        <p className="font-semibold text-4xl mx-5">{quantity}</p>
        <button id="add-item-btn" className="btn-transparent" onClick={addItemToCartHAndler}>
          <PlusIcon className="h-6 w-6" aria-hidden="true" />
        </button>
      </div>
      <div className="font-bold text-xl m-auto">
        <p>Total: {totalPrice.toFixed(2)}</p>
      </div>
    </div>
  );
}
export default CartItem;
