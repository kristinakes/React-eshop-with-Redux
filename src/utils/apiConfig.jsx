export const API_PRODUCTS_URL = "http://localhost:8000/products/";
export const API_CART_URL = "http://localhost:8000/cart/";