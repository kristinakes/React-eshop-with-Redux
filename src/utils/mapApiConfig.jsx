//Google maps API key to be used in contacts page. Key should be protected. Update accordingly.
///More info https://developers.google.com/maps/documentation/javascript/get-api-key
export const GOOGLE_API_KEY = "Add your API Key here";