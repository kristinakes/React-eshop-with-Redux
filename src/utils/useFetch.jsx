import { useState, useEffect } from "react";

function useFetch(url, options) {
  const [data, setData] = useState(null);
  const [isPending, setIsPending] = useState(true);
  const [error, setError] = useState(null);

  useEffect(()=> {
    setTimeout(()=> {
        fetch(url, options).then(result => {
            if (!result.ok) {
                throw Error('Error fetching data');
            }
            return result.json();
        }).then(data => {
            setData(data);
            setIsPending(false);
            setError(null);
        }).catch(err => {
            setIsPending(false);
            setError(err.message)
        });
    }, 1000);
  }, [url, options]);

  return {data, isPending, error};
}
export default useFetch;
