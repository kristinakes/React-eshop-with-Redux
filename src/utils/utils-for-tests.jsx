import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import cartSlice from "../store/cartSlice";
import uiSlice from "../store/uiSlice";
import productSlice from "../store/productSlice"

export function renderWithProviders(
  ui,
  {
    preloadedState = {},

    store = configureStore({
      reducer: {
        products: productSlice.reducer,
        cart: cartSlice.reducer,
        ui: uiSlice.reducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
