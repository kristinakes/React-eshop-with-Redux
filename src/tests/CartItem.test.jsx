import { renderWithProviders } from "../utils/utils-for-tests";
import CartItem from "../components/CartItem";
import { MemoryRouter } from "react-router-dom";

describe("CartItem", () => {
  const item = {
    id: 1,
    title: "laptop",
    quantity: 1,
    totalPrice: 10,
    year: 2020,
    ram: 5,
    warranty: 2,
    price: 10,
  };

  it("renders correctly when there is a single item in cart", () => {
    const view = renderWithProviders(
      <MemoryRouter>
        <CartItem item={item} />
      </MemoryRouter>
    );
    
    expect(view.baseElement).toMatchSnapshot();
  });
});
