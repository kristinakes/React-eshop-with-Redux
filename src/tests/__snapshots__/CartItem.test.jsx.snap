// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem renders correctly when there is a single item in cart 1`] = `
<body>
  <div>
    <div
      class="grid grid-cols-2 gap-4 min-w-fit justify-center content-center sm:grid-cols-4 sm:gap-2 shadow-md m-5 p-3"
    >
      <div
        class="size-48 m-auto"
      >
        <a
          href="/Product/1"
        >
          <img
            alt="laptop"
            class="object-fill scale-75 z-0"
            src="/assets/product_images/laptop.png"
          />
        </a>
      </div>
      <div>
        <a
          href="/Product/1"
        >
          <h3
            class="font-bold text-xl text-blue my-5"
          >
            laptop
          </h3>
        </a>
        <div
          class="font-semibold flex flex-wrap gap-3 md:flex-col md:gap-1"
        >
          <p>
            Year: 
            2020
          </p>
          <p>
            RAM memory: 
            5
          </p>
          <p>
            Warranty: 
            2
          </p>
          <p>
            Price: 
            10
          </p>
        </div>
      </div>
      <div
        class="flex justify-center items-center"
      >
        <button
          class="btn-blue"
          id="remove-item-btn"
        >
          <svg
            aria-hidden="true"
            class="h-6 w-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
        <p
          class="font-semibold text-4xl mx-5"
        >
          1
        </p>
        <button
          class="btn-transparent"
          id="add-item-btn"
        >
          <svg
            aria-hidden="true"
            class="h-6 w-6"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4.5v15m7.5-7.5h-15"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        class="font-bold text-xl m-auto"
      >
        <p>
          Total: 
          10.00
        </p>
      </div>
    </div>
  </div>
</body>
`;
